### PR TITLE
8265711: C1: Intrinsify Class.getModifier method

### PR DESCRIPTION
**Pull request by kelthuzadx**
It's relatively a common case to get modifiers from a constant Class instance, i.e. ThirdPartyClass.class.getModifiers(). Currently, C1 Canonicalizer missed the opportunity of replacing Class.getModifiers intrinsic calls with compile-time constants.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265711](https://bugs.openjdk.java.net/browse/JDK-8265711): C1: Intrinsify Class.getModifier method

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 5a0716c863aead4647e668e401a9ad6a9a410421
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 3ba69f6a8be14b8224cd943d2efbd408aa0fcc71

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3616/head:pull/3616` \
`$ git checkout pull/3616`

Update a local copy of the PR: \
`$ git checkout pull/3616` \
`$ git pull https://git.openjdk.java.net/jdk pull/3616/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3616`

View PR using the GUI difftool: \
`$ git pr show -t 3616`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3616.diff">https://git.openjdk.java.net/jdk/pull/3616.diff</a>

</details>

### CODE REVIEW

**Comment by kelthuzadx** [comment]
Can I have a review of this PR? Thanks!

File: test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java:94]: The thousand separator is at the wrong place, should be `10_000`.

File: src/hotspot/share/c1/c1_Canonicalizer.cpp
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_Canonicalizer.cpp:551]: Although old code below has this weird style, please change this to `_getModifiers: {`

File: src/hotspot/share/c1/c1_Canonicalizer.cpp
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_Canonicalizer.cpp:558]: Shouldn't `t->is_primitive_type()` be handled as well?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: src/hotspot/share/c1/c1_Canonicalizer.cpp
Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_Canonicalizer.cpp:558]: Done.

File: src/hotspot/share/c1/c1_LIRGenerator.cpp
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: The previous version of your patch was missing the check for a primitive type. Why did testing not catch this?

Also, your PR title is misleading. You are not simply canonicalizing `Class.getModifier.` You are intrinsifying it.

Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: Because is_intrinsic_support does not know the fact that getModifier is intrinsic, so GraphBuilder will not generate Intrinsic instruction, thus canonicalizer is even not applied.

>  Also, your PR title is misleading. You are not simply canonicalizing Class.getModifier. You are intrinsifying it.

Kind remind, it’s indeed mislead  now.

Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: Because is_intrinsic_support is not aware of the fact that getModifiers is an intrinsic method. So canonicalizer is even not applied.

> Also, your PR title is misleading. You are not simply canonicalizing Class.getModifier. You are intrinsifying it.

Yes, exactly. Changed.

Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: With the new IR test framework, these kinds of problem would never happen, we can do verification at IR level.

Comment (TobiHartmann) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: > Because is_intrinsic_support is not aware of the fact that getModifiers is an intrinsic method. So canonicalizer is even not applied.

But your first version (https://github.com/openjdk/jdk/pull/3616/commits/7f0be4e8a2cb4f6a9ef650b1b31c16f46716a257) already had the change to `is_intrinsic_supported`? Shouldn't that version have triggered test failures because it does not handle primitive types?

> With the new IR test framework, these kinds of problem would never happen, we can do verification at IR level.

Unfortunately not. The IR test framework currently only supports C2.

Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: > But your first version (7f0be4e) already had the change to is_intrinsic_supported? Shouldn't that version have triggered test failures because it does not handle primitive types?

The first version only changes Canonicalizer::do_Intrinsic :_

Comment (TobiHartmann) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: Okay, maybe the force-push screwed up some of the history here, you should try to avoid that (it also makes it much harder to review changes incrementally).

File: src/hotspot/share/c1/c1_LIRGenerator.cpp
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_LIRGenerator.cpp:1330]: `temp` could be renamed to `klass`.

File: src/hotspot/share/c1/c1_LIRGenerator.cpp
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_LIRGenerator.cpp:1339]: Why are you reloading the klass here?

Also, as @rwestrel explained in the RFR for [8211231](https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2018-September/030745.html), we need to be really careful when emitting branches in C1's LIR.

File: src/hotspot/share/c1/c1_LIRGenerator.cpp
Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: > Okay, maybe the force-push screwed up some of the history here, you should try to avoid that (it also makes it much harder to review changes incrementally).

Got it! (The reason why I forcibly push is that I can not push them normally after merging some commits from master branch.)

Comment (TobiHartmann) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1333]: Even when merging changes from master into your branch, a force-push should not be required. Here's a related discussion:
https://mail.openjdk.java.net/pipermail/hotspot-dev/2020-November/046538.html

File: src/hotspot/share/c1/c1_LIRGenerator.cpp
Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_LIRGenerator.cpp:1339]: > Why are you reloading the klass here?

Typo. I forget removing it.

> Also, as @rwestrel explained in the RFR for 8211231, we need to be really careful when emitting branches in C1's LIR.

> The problem is that there's control flow here that's invisible to
the register allocator. The register allocator only knows about control
flow that connects basic blocks. The control flow here is within a
block. So as far as the register allocator is concerned the code above
is a linear sequence of instructions all of which are all executed

Ok, I see. But from the compiled code, it behaves well：

```
B1 [0, 0] sux: B0 
__id_Instruction___________________________________________
   0 label [label:0x00007f0c10025f60]
   2 std_entry 

B0 std [0, 4] preds: B1 
__id_Instruction___________________________________________
   8 label [label:0x00007f0c100253c0]
  10 move [Base:[rsi|L] Disp: 16|] [rax|M]  [bci:1]
  12 cmp [NE] [rax|M] [metadata:0x0000000000000000|M] 
  14 branch [NE] [label:0x00007f0c10028d78] 
  16 move [int:1041|I] [rsi|I] 
  18 branch [AL] [label:0x00007f0c10028e18] 
  20 label [label:0x00007f0c10028d78]
  22 move [Base:[rax|M] Disp: 168|I] [rsi|I] 
  24 label [label:0x00007f0c10028e18]
  26 move [rsi|I] [rax|I] 
  28 return [rax|I]  

```

```

__bci__use__tid____instr____________________________________
. 0    0     4     B1 [0, 0] -> B0 sux: B0

   0 label [label:0x00007f0c10025f60]
Memory range [0x00007f0ca06a9120..0x00007f0ca06a9120] not readable   2 std_entry 
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a9120:   mov    %eax,-0x16000(%rsp)
  0x00007f0ca06a9127:   push   %rbp
  0x00007f0ca06a9128:   sub    $0x30,%rsp
--------------------------------------------------------------------------------
__bci__use__tid____instr____________________________________
. 0    0     0     B0 (SV) [0, 4] dom B1 pred: B1

   8 label [label:0x00007f0c100253c0]
Memory range [0x00007f0ca06a912c..0x00007f0ca06a912c] not readable  10 move [Base:[rsi|L] Disp: 16|] [rax|M]  [bci:1]
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a912c:   mov    0x10(%rsi),%rax
--------------------------------------------------------------------------------
  12 cmp [NE] [rax|M] [metadata:0x0000000000000000|M] 
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a9130:   cmp    $0x0,%rax
--------------------------------------------------------------------------------
  14 branch [NE] [label:0x00007f0c10028d78] 
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a9134:   jne    0x00007f0ca06a913a
--------------------------------------------------------------------------------
  16 move [int:1041|I] [rsi|I] 
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a913a:   mov    $0x411,%esi
--------------------------------------------------------------------------------
  18 branch [AL] [label:0x00007f0c10028e18] 
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a913f:   jmpq   0x00007f0ca06a9144
--------------------------------------------------------------------------------
  20 label [label:0x00007f0c10028d78]
Memory range [0x00007f0ca06a9144..0x00007f0ca06a9144] not readable  22 move [Base:[rax|M] Disp: 168|I] [rsi|I] 
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a9144:   mov    0xa8(%rax),%esi
--------------------------------------------------------------------------------
  24 label [label:0x00007f0c10028e18]
Memory range [0x00007f0ca06a914a..0x00007f0ca06a914a] not readable  26 move [rsi|I] [rax|I] 
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a914a:   mov    %rsi,%rax
--------------------------------------------------------------------------------
  28 return [rax|I]  
--------------------------------------------------------------------------------
 ;;  block B1 [0, 0]
  0x00007f0ca06a914d:   add    $0x30,%rsp
  0x00007f0ca06a9151:   pop    %rbp
  0x00007f0ca06a9152:   cmp    0x118(%r15),%rsp
 ;;  block B0 [0, 4]
  0x00007f0ca06a9159:   ja     0x00007f0ca06a915f
  0x00007f0ca06a915f:   retq   
--------------------------------------------------------------------------------
```
Also tests under compiler and jdk are passed with slowdebug mode.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by kelthuzadx** [comment]
Thank Tobias for reviews!

**Comment by kelthuzadx** [comment]
Can I have a second review of this PR？

Thanks,
Yang

**Comment by kelthuzadx** [comment]
PING：Can I have a second review of this PR？

File: test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java:31]: I would suggest to add 2 additional `@run` command to make sure test passed in all modes:
 - default: without `-XX:TieredStopAtLevel`
 - C2 only: with `-XX:-TieredCompilation`

Comment (kelthuzadx) [reply test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java:31]: Thank you Vladimir for taking time to look at this. Good suggestion! I've added.

File: test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java:52]: `-XX:TieredStopAtLevel=4` does not work when Tiered is off. You don'ty need this flag.

**Review by vnkozlov — APPROVED** [review]
Just one small comment you can fix before push. Otherwise it is good.

**Comment by kelthuzadx** [comment]
Okay, I learned this from [valhalla project](https://github.com/openjdk/valhalla/blob/c8d7c8260921ccb2b2e72dcc6cc04144ff7649fc/test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1.java#L61). May be it's redundant if `-XX:-TieredCompilation` is disabled.

File: test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java
Comment (kelthuzadx) [reply test/hotspot/jtreg/compiler/c1/CanonicalizeGetModifiers.java:52]: Okay, I learned this from [Valhalla project](https://github.com/openjdk/valhalla/blob/c8d7c8260921ccb2b2e72dcc6cc04144ff7649fc/test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestC1.java#L61). Maybe it's redundant if -XX:-TieredCompilation is disabled.

**Comment by kelthuzadx** [comment]
Thanks Tobias and Vladimir for reviews!

/integrate

**Pull request closed without merging**
